Documentation/design-docs/COM-activation.md

```diff
@@ -0,0 +1,105 @@
+
+### .NET Core Class COM Activation
+
+In .NET Core, our intent will be to avoid a system wide shim library. This may add additional cost for deployment scenarios, but will reduce servicing and engineering costs by making deployment more explicit and less magic.
+
+The current hosting solutions are defined at [Documentation/design-docs/host-components.md](https://github.com/dotnet/core-setup/blob/master/Documentation/design-docs/host-components.md). Along with the existing hosts an additional COM activation host library will be added. This library will export the required functions for COM class activation.
+
+>[`HRESULT DllGetClassObject(REFCLSID rclsid, REFIID riid, LPVOID *ppv);`](https://docs.microsoft.com/en-us/windows/desktop/api/combaseapi/nf-combaseapi-dllgetclassobject)
+
+>[`HRESULT DllCanUnloadNow();`](https://docs.microsoft.com/en-us/windows/desktop/api/combaseapi/nf-combaseapi-dllcanunloadnow)
+
+When `DllGetClassObject()` is called in an activation scenario, the following will occur:
+
+1) Determine additional registration information needed for activation.
+    * **The location and mechanics of accessing this registration information is still TBD**
+    * `Assembly` (**required**) - the [Fully-Qualified Name](https://docs.microsoft.com/en-us/dotnet/framework/reflection-and-codedom/specifying-fully-qualified-type-names) of the assembly.
+    * `Class` (**required**) - the full type name (e.g. `MyCompany.MyProduct.MyClass`).
+    * `Codebase` (**required**) - an absolute path to the assembly to load. If a [`runtimeconfig.json`](https://github.com/dotnet/cli/blob/master/Documentation/specs/runtime-configuration-file.md) file exists adjacent to the assembly that file will be used to describe runtime configuration details.
+    * `RuntimeVersion` (optional) - if not supplied the latest CLR found will be used.
+1) Using the existing `hostfxr` library, attempt to discover the desired CLR.
+    * A current running CLR will be preferred, otherwise a new CLR instance will be created.
+1) A request to the CLR will be made via a new method for class activation within a COM environment.
+    * The ability to load the assembly and then create a class instance will require a function be exposed that can be called from `hostfxr`.
+    * Example of API in `System.Private.CoreLib` on the `Activator` class:
+        ``` csharp
+        public static class Activator
+        {
+            ...
+            #if WINDOWS
+            public static ObjectHandle CreateInstanceForCom(string assemblyName, string typeName);
+            #endif
+            ...
+        }
+        ```
+        Note this API would not be exposed outside of `System.Private.CoreLib`.
+    * The loading of the assembly will take place in a new [`AssemblyLoadContext`](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.loader.assemblyloadcontext?view=netcore-2.1) for dependency isolation.
```

Comment:
This will require much more detail eventually. Things which feel like being problematic:
- What type of resolution logic will the new ALC use? (The potentially upcoming .deps.json based ALC, or something else).
- Do we need / will there be a way to customize the ALC used here.
- How much do we share - how much we enforce isolation in ALCs versus sharing through default.
- What happens if the same assembly is requested twice - do we load it again into a new ALC - or do we maintain a map of COM-loaded assemblies.
- Thinking about scenarios like: Creating several instances of some COM object, with the expectation that they have a way to talk to each other.

Reply:
This is a good list of concerns. Here are my initial thoughts:

>What type of resolution logic will the new ALC use? (The potentially upcoming .deps.json based ALC, or something else).

I am inclined to use whatever would be used if the server was used as a console application, so yes?

>Do we need / will there be a way to customize the ALC used here.

I think that is the `.deps.json` or `.runtimeconfig.json`. I think extending the ALC beyond the `*.json`s for COM activation is a tricky situation and something we should avoid.

>How much do we share - how much we enforce isolation in ALCs versus sharing through default.

I would prefer to isolate as much as possible. This opinion is based on trying to reduce situations where users can get themselves into intractable problems. However that opinion may be tempered based on what @richlander has to say.

>What happens if the same assembly is requested twice - do we load it again into a new ALC - or do we maintain a map of COM-loaded assemblies.

I am inclined to match the native server behavior and once the assembly is loaded that context is used for all further class activation from that assembly.

>Thinking about scenarios like: Creating several instances of some COM object, with the expectation that they have a way to talk to each other.

Another great question for @richlander. This is a real scenario that needs to be investigated.